### PR TITLE
Add Windows support to cacheddownloader

### PR DESCRIPTION
**Pull request by lvarvel**
Cacheddownloader has a number of issues on Windows, mostly due to irritating differences between Windows and Linux file open/closing and removal. This fixes one of them.

There's no additional test code necessary, because this change actually fixes three tests that currently fail when run on Windows.

After this change, there will still be 29 failing tests (previously 32), and we'll be submitting more pull requests as we stomp those.

### CODE REVIEW

**Comment by ematpl** [comment]
Thanks, @lvarvel! I've created [this Diego story](https://www.pivotaltracker.com/story/show/88834310) to evaluate the PR.

**Comment by krishicks** [comment]
Could you please clarify which tests were failing on Windows? Was it the cacheddownloader unit tests, or other tests from diego-release?

Thanks,
@krishicks && @ematpl

**Comment by lvarvel** [comment]
The tests were the cacheddownloader unit tests.  I don't think we had any other failing diego tests.

**Comment by lvarvel** [comment]
There is also one final test that fails intermittently on Windows. We've dropped a chore into our backlog, and we'll get back to you on that one later.  This PR should be good to go.

**Comment by ematpl** [comment]
@lvarvel @mavenraven: The first commit was fine, but the second one (with the `windowsSafeRename` function) is failing on the Travis linux build. Could you please remove that commit and submit a separate PR for that one? Also, the Travis build is failing because that commit depends on several `syscall` functions that are unavailable on Linux. It may be possible to put the architecture-dependent details into separate files with different go build tags for the different platforms to get this to compile correctly, but it would be far preferable for us to find a solution that doesn't rely on these low-level details.

Thanks,
@ematpl && @matt-royal

**Comment by lvarvel** [comment]
Sorry about that.  We've made some changes to use build tags.

**Comment by lvarvel** [comment]
Related: https://github.com/golang/go/issues/8914

TLDR: os.Rename doesn't work as expected on Windows, and the Golang core team won't fix it.  They *might* add an os.Replace function in go 1.5 that does the right thing (essentially what we've done here).

**Comment by ematpl** [comment]
Excellent, thanks for the reference to the outstanding issue. If we can remove this with Go 1.5 or later, that would be great.

We tried the entire units/inigo/DATs run on OS X/bosh-lite and it seems fine, so we can merge and push this once ketchup is green. In the meantime, I have a few minor comments to make inline.

Thanks,
Eric

File: cached_downloader.go [hunk 1]
```diff
@@ -22,8 +22,8 @@ type CachedDownloader interface {
 	Fetch(urlToFetch *url.URL, cacheKey string, transformer CacheTransformer, cancelChan <-chan struct{}) (io.ReadCloser, int64, error)
 }
 
-func NoopTransform(source, destination string) (int64, error) {
-	err := os.Rename(source, destination)
+func NoopTransform(source string, destination string) (int64, error) {
```
Comment (ematpl) [inline cached_downloader.go]: It's idiomatic Go to collapse type declarations for adjacent arguments of the same type, as was done before the change.

File: replace.go [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// +build !windows
+
+package cacheddownloader
+
+import "os"
+
+func Replace(src, dst string) error {
```
Comment (ematpl) [inline replace.go]: Let's not export the `Replace` function from the package unless we actually need it elsewhere.

File: replace_windows.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	if err != nil {
+		return err
+	}
+	defer syscall.FreeLibrary(kernel32)
+	moveFileExUnicode, err := syscall.GetProcAddress(kernel32, "MoveFileExW")
+	if err != nil {
+		return err
+	}
+
+	srcString := syscall.StringToUTF16Ptr(src)
```
Comment (ematpl) [inline replace_windows.go]: According to https://golang.org/src/syscall/syscall_windows.go, `StringToUTF16Ptr` is deprecated in favor of `UTF16PtrFromString` (which returns an error instead of panicking on certain error cases).

**Comment by mavenraven** [comment]
We've completed the requested changes, waiting for Varvel to give us push access.

**Comment by lvarvel** [comment]
Access granted, mavenraven.

We've completed the requested changes, waiting for Varvel to give us push
access.

—
Reply to this email directly or view it on GitHub
<https://github.com/pivotal-golang/cacheddownloader/pull/6#issuecomment-75969499>
.
